### ISSUE TITLE
Add Keyboard Accessibility and Conduct Keyboard Testing

### ISSUE THREAD

**Issue by kumokuro** [TheOdinProject/theodinproject#2893]
Complete the following REQUIRED checkboxes:

-   [x ] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [ x] The title of this issue follows the `brief description of request` format, e.g. `Add dark mode to website`

The following checkbox is OPTIONAL:

-   [ x] I would like to be assigned this issue to work on it 

<hr>

**1. Description of the Feature Request:**

### **Feature Request** 
I suggest that The Odin Project works to ensure that all functionality of content throughout the site is able to be accessed and navigable by keyboard, in line with guidelines set by the [WCAG](https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0&showtechniques=211#qr-keyboard-operation-keyboard-operable). In addition, [keyboard testing should be conducted](https://webaccess.berkeley.edu/resources/tips-and-how-tos/how-do-keyboard-testing) to identify the functions on each page of the site, and determine whether these functions are able to be accessed by keyboard alone. 

### **What is Keyboard Accessibility?**
Keyboard accessibility is one of the most important aspects of web accessibility. Many users with motor disabilities rely on a keyboard. Some people have tremors which don't allow for fine muscle control. Others have little or no use of their hands, or no hands at all. In addition to traditional keyboards, some users may use modified keyboards or other hardware that mimics the functionality of a keyboard. Blind users also typically use a keyboard for navigation. Users without disabilities may use a keyboard for navigation because of preference or efficiency.

**2. Acceptance Criteria:**
- [ ] All functionality on each page has been identified and tested for keyboard accessibility 
- [ ] All elements that can be clicked with a mouse are focusable using the keyboard
- [ ] Focusable elements have interactive semantics (ex: is a link or changes an option) 
- [ ] Interactive elements are able to be activated with a keyboard
- [ ] Focusable elements have focus styling indicating that the element is focused 
- [ ] Keyboard tab order is coordinated with the structure of the webpage 
- [ ] All pages are free of keyboard traps (cannot move onto the next element using tab or escape with the escape key) 

**3. Additional Information:**

### **For More on Keyboard Accessibility:** 
- [Web Content Accessibility Guidelines on Keyboard Accessibility](http://www.w3.org/TR/WCAG20/#keyboard-operation-keyboard-operable)
- [Designing for Keyboard Accessibility](https://www.washington.edu/accessibility/checklist/keyboard/)
- [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Keyboard)  
- [Web Accessibility Criteria - Keyboard Accessibility](https://www.csun.edu/universal-design-center/web-accessibility-criteria-keyboard-accessibility)

**Comment by KevinMulhern** [comment]
Thanks for taking the time to outline this @kumokuro 💪 I don't believe there should be too much work involved in this. We don't have that many forms fortunately. 

But we're converting all the CSS on the site over to Tailwind at the moment so there is quite a bit of churn happening with the front end of the app which would conflict with this if it is worked on at the same time. Once the Tailwind project is complete we can open this up to be worked on.

**Comment by KevinMulhern** [comment]
This ones available to work on! first step would be identifying specific areas that can be improved and then breaking those out into individual issues.

**Comment by dperry83** [comment]
@KevinMulhern @ChargrilledChook  If this one is still up, I'd like to be assigned!

**Comment by KevinMulhern** [comment]
Thanks @dperry83, let us know if we can help with anything!

**Comment by dsmchen** [comment]
Hi @KevinMulhern & @dperry83, I notice this issue has label "Status: Help Wanted". I'm happy to help! :)

**Comment by dsmchen** [comment]
Hi @KevinMulhern @dperry83 @ChargrilledChook - I'd like to follow up on the above. 🙏

**Comment by KevinMulhern** [comment]
Sorry for missing your first comment @dsmchen, this hasn't had any activity from it's assignee so I'll reassign it to you!

Are you happy with the approach of breaking it down into the specific pages/areas to test and improve?

**Comment by dsmchen** [comment]
@KevinMulhern - No worries! I'm happy with the approach. I'm thinking we start with the top 10 most-visited pages. What do you think?

**Comment by KevinMulhern** [comment]
@dsmchen top 10 most-visited pages will be mostly lessons, which could be misleading since all lessons use the same page/template within the app.

I think we'd likely be better focusing on features/flows on the site where keyboard accessibility is most important. 

- The authentication flows: login/sign up/password reset pages.
- The curriculum flow: path -> course -> lesson.

**Comment by dsmchen** [comment]
@KevinMulhern - Good point, I agree

I created a gist to define scope and list bugs: https://gist.github.com/dsmchen/68d1f5a70a6ca0567ee4b6e660455169

Should I create an issue for each bug?

**Comment by dsmchen** [comment]
Hi @KevinMulhern, in case you missed the above 🙏

**Comment by KevinMulhern** [comment]
Sorry for the delay @dsmchen, and thanks for putting that Gist together 💪

I'd like to get input from our more accessibility minded maintainers: CC: @thatblindgeye, @mao-sz and any other @TheOdinProject/maintainers that are interested in this - heres the list of proposed changes: https://gist.github.com/dsmchen/68d1f5a70a6ca0567ee4b6e660455169

----
> Bug: Add "Skip to main content" link

TIL this is a thing :) I think I'm better leaving this to another maintainer to make a decision on because of my lack of experience with it.

> Bug: Fix header logo anchor element

Can you explain this one a little more please @dsmchen? It reads like we'd allow users to toggle the header logo to be hidden or displayed?

> Bug: Add focus style to GitHub and Google auth buttons

Nice catch, easy yes to this!

> Bug: Fix focus style

Aesthetically I think it looks better to have the focus outline the same as the button background color. But... if it's causing a big accessibility issue I wouldn't be against us changing it. I guess what I am really asking is how big of an issue is this one and if it's worth changing?  

> Bug: Fix "Lesson contents" focus order

Nice catch, this was raised in a [recent issue](https://github.com/TheOdinProject/theodinproject/issues/4423) and there were multiple solutions raised. I think we should leave this one to that other issue.

**Comment by dsmchen** [comment]
Hey @KevinMulhern, no worries!

---

>> Bug: Fix header logo anchor element

> Can you explain this one a little more please @dsmchen? It reads like we'd allow users to toggle the header logo to be hidden or displayed?

<img width="1728" height="492" alt="Image" src="https://github.com/user-attachments/assets/a53a5bde-4aba-4710-bd08-8c82de3b7c8a" />

Current: On desktop, `a.logo` is shown and focusable (small blue dot in the top-left corner)
Expected: On desktop, `a.logo` is hidden and therefore not focusable

Current: On mobile, the desktop logo anchor element is shown
Expected: On mobile, the desktop logo anchor element is hidden

>> Bug: Fix focus style

> Aesthetically I think it looks better to have the focus outline the same as the button background color. But... if it's causing a big accessibility issue I wouldn't be against us changing it. I guess what I am really asking is how big of an issue is this one and if it's worth changing?

The main issue is that the contrast ratio of the secondary button focus appearance in dark mode is approx 2:1, while the success criterion is at least 3:1. For more info: https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance.

**Comment by dsmchen** [comment]
Hey @thatblindgeye and @mao-sz, in case you missed @KevinMulhern's comment: https://github.com/TheOdinProject/theodinproject/issues/2893#issuecomment-3568471905 🙏

**Comment by mao-sz** [comment]
I think Eric will have more helpful insight on focus order stuff.

Regarding the logo anchor, is there a reason we have 2 anchors for different breakpoints? The anchors are functionally identical (same href and aria-label), the only different being the contents. I'd say we axe the first logo and move the `<img>` inside the second one. We can also drop `block` from the `<img>` classes since that's the default anyway.

<img width="893" height="78" alt="Image" src="https://github.com/user-attachments/assets/00c4cc6d-0259-42c5-ad85-21060e4714db" />

---

The sign-in button focus styles do exist, they just need to have the colours tweaked as they currently apply the exact same colour, hence no visible change.

---

I'm not sure we need to keep the focus style colours consistent. The shape being the same (outline) should be sufficient, and as long as they're actually visible enough (unlike the sign-in buttons pointed out!). If the issue is just some colour contrasts aren't at least 3:1, then we can definitely tweak the relevant colours to meet that ratio.

**Comment by dsmchen** [comment]
@mao-sz - Thanks for the feedback!

@thatblindgeye - In case you missed this!

**Comment by thatblindgeye** [comment]
What @mao-sz said makes sense to me, and yeah there's other issues related to the focus order of stuff so agreed that part can be tackled in that other issue rather that as part of this.

**Comment by dsmchen** [comment]
Hey @KevinMulhern, @mao-sz & @thatblindgeye - I've updated the gist to reflect our discussion: https://gist.github.com/dsmchen/68d1f5a70a6ca0567ee4b6e660455169

Shall I create an issue for each bug?

**Comment by mao-sz** [comment]
Go ahead and make separate issues for each bug; we may need to discuss specifics for implementation for some. I'll set this issue as their parent later.

**Comment by dsmchen** [comment]
Hey @mao-sz, here are the issues:
- https://github.com/TheOdinProject/theodinproject/issues/5199
- https://github.com/TheOdinProject/theodinproject/issues/5200
- https://github.com/TheOdinProject/theodinproject/issues/5201
- https://github.com/TheOdinProject/theodinproject/issues/5202
